File: pkg/api/memory.go
package orla

import "context"

// CachePolicy constants for stage-level cache policy overrides.
const (
	CachePolicyPreserve = "preserve"
	CachePolicyFlush    = "flush"
	CachePolicyAuto     = "auto"
)

// CacheHints are optional per-stage parameters that influence the Memory Manager's
// cache decisions. They follow the same pattern as SchedulingHints.
type CacheHints struct {
	PreserveThresholdTokens *int  `json:"preserve_threshold_tokens,omitempty"`
	FlushOnComplete         *bool `json:"flush_on_complete,omitempty"`
}

// CacheEvent describes a stage transition for the MemoryPolicy to evaluate.
type CacheEvent struct {
	PrevStageBackend string
	PrevStageModel   string
	NextStageBackend string
	NextStageModel   string
	DeltaTokens      int
	TotalTokens      int
	TransitionType   string // "stage", "agent", "workflow_complete"
}

// MemoryPolicy determines cache actions at workflow level. Developers can
// implement custom policies or use the built-in ones shipped with Orla.
type MemoryPolicy interface {
	Decide(ctx context.Context, event CacheEvent) string // returns CachePolicyPreserve, CachePolicyFlush, or CachePolicyAuto
}

// --- Built-in policies ---

// preserveOnSmallIncrementPolicy preserves cache when the context delta
// is below a threshold and the backend/model hasn't changed.
type preserveOnSmallIncrementPolicy struct {
	thresholdTokens int
}

// NewPreserveOnSmallIncrementPolicy creates a MemoryPolicy that preserves
// KV cache when the new stage adds fewer than thresholdTokens to the context.
func NewPreserveOnSmallIncrementPolicy(thresholdTokens int) MemoryPolicy {
	if thresholdTokens <= 0 {
		thresholdTokens = 256
	}
	return &preserveOnSmallIncrementPolicy{thresholdTokens: thresholdTokens}
}

func (p *preserveOnSmallIncrementPolicy) Decide(_ context.Context, event CacheEvent) string {
	if event.PrevStageBackend != event.NextStageBackend || event.PrevStageModel != event.NextStageModel {
		return CachePolicyAuto
	}
	if event.PrevStageBackend == "" {
		return CachePolicyAuto
	}
	if event.DeltaTokens <= p.thresholdTokens {
		return CachePolicyPreserve
	}
	return CachePolicyAuto
}

// flushAtBoundaryPolicy flushes cache at workflow completion or backend/model switches.
type flushAtBoundaryPolicy struct{}

// NewFlushAtBoundaryPolicy creates a MemoryPolicy that flushes cache at
// workflow boundaries and when the backend/model changes between stages.
func NewFlushAtBoundaryPolicy() MemoryPolicy {
	return &flushAtBoundaryPolicy{}
}

func (p *flushAtBoundaryPolicy) Decide(_ context.Context, event CacheEvent) string {
	if event.TransitionType == "workflow_complete" {
		return CachePolicyFlush
	}
	if event.PrevStageBackend != "" &&
		(event.PrevStageBackend != event.NextStageBackend || event.PrevStageModel != event.NextStageModel) {
		return CachePolicyFlush
	}
	return CachePolicyAuto
}

// composedMemoryPolicy chains multiple policies; first non-auto result wins.
type composedMemoryPolicy struct {
	policies []MemoryPolicy
}

// MemoryPolicyOption configures a DefaultMemoryPolicy.
type MemoryPolicyOption func(*defaultMemoryPolicyConfig)

type defaultMemoryPolicyConfig struct {
	preserveThreshold int
}

// WithPreserveThreshold overrides the default token threshold for the
// preserve-on-small-increment policy.
func WithPreserveThreshold(tokens int) MemoryPolicyOption {
	return func(c *defaultMemoryPolicyConfig) {
		c.preserveThreshold = tokens
	}
}

// NewDefaultMemoryPolicy creates a MemoryPolicy that composes the three
// paper policies: preserve on small increment, then flush at boundary.
// The flush-under-pressure policy is handled server-side by the Memory Manager.
func NewDefaultMemoryPolicy(opts ...MemoryPolicyOption) MemoryPolicy {
	cfg := &defaultMemoryPolicyConfig{preserveThreshold: 256}
	for _, opt := range opts {
		opt(cfg)
	}
	return &composedMemoryPolicy{
		policies: []MemoryPolicy{
			NewPreserveOnSmallIncrementPolicy(cfg.preserveThreshold),
			NewFlushAtBoundaryPolicy(),
		},
	}
}

func (p *composedMemoryPolicy) Decide(ctx context.Context, event CacheEvent) string {
	for _, policy := range p.policies {
		result := policy.Decide(ctx, event)
		if result != CachePolicyAuto {
			return result
		}
	}
	return CachePolicyAuto
}

File: pyorla/src/pyorla/memory.py
"""Memory and KV-cache policy types for multi-stage workflows."""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass

from pyorla.types import CACHE_POLICY_AUTO, CACHE_POLICY_FLUSH, CACHE_POLICY_PRESERVE


@dataclass
class CacheEvent:
    """Describes a stage transition for the MemoryPolicy to evaluate."""

    prev_stage_backend: str = ""
    prev_stage_model: str = ""
    next_stage_backend: str = ""
    next_stage_model: str = ""
    delta_tokens: int = 0
    total_tokens: int = 0
    transition_type: str = ""  # "stage" | "workflow_complete"


class MemoryPolicy(ABC):
    """Determines cache actions at workflow level."""

    @abstractmethod
    def decide(self, event: CacheEvent) -> str:
        """Return one of CACHE_POLICY_PRESERVE, CACHE_POLICY_FLUSH, CACHE_POLICY_AUTO."""
        ...


class PreserveOnSmallIncrementPolicy(MemoryPolicy):
    """Preserve KV cache when context delta is small and backend hasn't changed."""

    def __init__(self, threshold_tokens: int = 256) -> None:
        self.threshold_tokens = max(threshold_tokens, 1)

    def decide(self, event: CacheEvent) -> str:
        if (
            event.prev_stage_backend != event.next_stage_backend
            or event.prev_stage_model != event.next_stage_model
        ):
            return CACHE_POLICY_AUTO
        if not event.prev_stage_backend:
            return CACHE_POLICY_AUTO
        if event.delta_tokens <= self.threshold_tokens:
            return CACHE_POLICY_PRESERVE
        return CACHE_POLICY_AUTO


class FlushAtBoundaryPolicy(MemoryPolicy):
    """Flush cache at workflow boundaries and backend switches."""

    def decide(self, event: CacheEvent) -> str:
        if event.transition_type == "workflow_complete":
            return CACHE_POLICY_FLUSH
        if event.prev_stage_backend and (
            event.prev_stage_backend != event.next_stage_backend
            or event.prev_stage_model != event.next_stage_model
        ):
            return CACHE_POLICY_FLUSH
        return CACHE_POLICY_AUTO


class DefaultMemoryPolicy(MemoryPolicy):
    """Composed policy: preserve-on-small-increment then flush-at-boundary."""

    def __init__(self, preserve_threshold: int = 256) -> None:
        self._policies: list[MemoryPolicy] = [
            PreserveOnSmallIncrementPolicy(preserve_threshold),
            FlushAtBoundaryPolicy(),
        ]

    def decide(self, event: CacheEvent) -> str:
        for policy in self._policies:
            result = policy.decide(event)
            if result != CACHE_POLICY_AUTO:
                return result
        return CACHE_POLICY_AUTO
